File: UFBA/ENGG03/engg03/eda.go
function adicionar_fila(fila, novo_nó):
    if fila->inicio == λ then
        fila->inicio = novo_nó
    else:
        if novo_nó->prioridade then
            atual = fila->inicio
            if atual == λ || !atual->prioridade then
                novo_nó->proximo = fila->inicio
                fila->inicio = novo_nó
            else
                while atual->proximo != && λ atual->proximo->prioridade do
                    atual = atual->proximo
                novo_nó->proximo = atual->proximo
                atual->proximo = novo_nó
        else
            atual = fila->inicio
            while atual->proximo =! λ do
                atual = atual->proximo
            atual->proximo = novo_nó
    return novo_nó

function remover_fila(fila):
    if fila->inicio == λ then
        print("Fila está vazia")
        return λ
    else:
        removido = fila->inicio
        fila->inicio = fila->inicio->proximo
        return removido

new(fila)
fila->inicio = λ

bool sair = false
while !sair do
    print("Selecione uma opção:")
    print("1 - Adicionar à fila")
    print("2 - Remover da fila")
    print("3 - Sair")
    opção = input()

    if opção == 1 then
        print("Adicionando à fila")
        new(nó)

        print("Nome: ")
        string nome = input()
        nó->nome = nome

        print("Prioridade (s/N): ")
        bool prioridade = input() == "s"
        nó->prioridade = prioridade

        string senha = gerar_senha()
        nó->senha = senha
        print("Guarde a senha:", senha)

        nó->proximo = λ
        adicionar_fila(fila, nó)
    else if opção == 2:
        print("Removendo da fila")
        nó = remover_fila(fila)
        if nó != λ:
            print("Nó removido:")
            print("Nome: " + nó->nome + ", Senha: " + nó->senha)
            delete(nó)

    else if opção == 3:
        sair = true

delete(fila)

File: UFBA/ENGG03/engg03/eda.py
import random
import string


# Gerar uma senha aleatória
def gerar_senha():
    return "".join(random.choices(string.ascii_uppercase, k=6))


def criar_nó(nome: str, prioridade: bool):
    return {
        "nome": nome,
        "senha": gerar_senha(),
        "prioridade": prioridade,
        "proximo": None,
    }


def criar_fila():
    return {"proximo": None, "prioridade": False}


def adicionar_fila(fila, novo_nó):
    if fila["proximo"] is None:
        fila["proximo"] = novo_nó
        return novo_nó

    if fila["proximo"]["prioridade"] is False and novo_nó["prioridade"] is True:
        novo_nó["proximo"] = fila["proximo"]
        fila["proximo"] = novo_nó
        return novo_nó

    if novo_nó["prioridade"] is False:
        atual = fila["proximo"]
        while atual["proximo"] is not None:
            atual = atual["proximo"]
        atual["proximo"] = novo_nó
        return novo_nó

    # PP NN P NN
    if novo_nó["prioridade"] is True:
        conta_comum = 0
        conta_prioridade = 0

        atual = fila["proximo"]

        while atual["proximo"] is not None:
            if atual["prioridade"] is False:
                conta_comum += 1
                conta_prioridade = 0
            else:
                conta_prioridade += 1
                conta_comum = 0
            atual = atual["proximo"]

            if conta_comum == 2 or conta_prioridade == 1:
                break
            if conta_comum == 1 or conta_prioridade == 1:
                break

        novo_nó["proximo"] = atual["proximo"]
        atual["proximo"] = novo_nó

    return novo_nó


def remover_fila(fila):
    if fila["proximo"] is None:
        print("Fila está vazia")
        return None
    else:
        removido = fila["proximo"]
        fila["proximo"] = fila["proximo"]["proximo"]
        return removido


fila = criar_fila()

adicionar_fila(fila, criar_nó("3", False))
adicionar_fila(fila, criar_nó("4", False))
adicionar_fila(fila, criar_nó("1", True))
adicionar_fila(fila, criar_nó("2", True))
adicionar_fila(fila, criar_nó("7", False))
adicionar_fila(fila, criar_nó("8", False))
adicionar_fila(fila, criar_nó("5", True))
adicionar_fila(fila, criar_nó("6", True))

sair = False
for _ in range(8):
    nó = remover_fila(fila)
    if nó is not None:
        print("Nó removido:")
        print(f"Nome: {nó['nome']}, Senha: {nó['senha']}")

# s s n n s s n n
# 1 2 3 4 5 6 7 8
